File: python_lib/src/selection_sort.py
from typing import List

def find_smallest(arr:List[int]) -> int:
    """
    This function finds the smallest number and return it's index

    Params:
        arr: a list of unsorted integers

    Returns:
        The index of the smallest integer in the arr
    """

    # getting the first element of the arr
    smallest = arr[0]
    # settin the default smallest index
    smallest_index = 0

    # looping ...
    for i in range(1, len(arr)):

        # checking if the current arr element is less than the previous smallest 
        if arr[i] < smallest:
            # if true
            smallest = arr[i]
            smallest_index = i
    return smallest_index



def selection_sort(arr:List[int]) -> List[int]:
    """
    This function takes an unsorted list

    Params:
        arr: a list of unsorted integers

    Returns:
        A sorted list
    """

    # performing type checks on the arr
    if not all( isinstance(elem , int) for elem in arr):
            raise TypeError("Only integer type is accepted as a valid arr element")
    
    copied_arr = list(arr)
    sorted_arr = []

    # looping ...
    for _ in range(len(copied_arr)):
        # getting the smallest index in the arr
        smallest_index = find_smallest(copied_arr)
        # updating the sorted arr
        sorted_arr.append(copied_arr.pop(smallest_index))

    return sorted_arr

File: typescript_lib/src/selectionSort.ts

/**
 * 
 * @param arr : an unsorted array
 * @returns : the index of the smallest element
 */
function findSmallest(arr: number[]): number{
    
    // setting default values
    let smallest = arr[0] as number
    let smallestIndex = 0

    // looping through the arr starting from the second element
    for (let i = 1; i < arr.length; i++){
        const currentElem = arr[i] as number
        if (currentElem < smallest) {
            smallest = currentElem
            smallestIndex = i
        }
    }

    // returning the 
    return smallestIndex 
}

/**
 * 
 * @param arr : an unsorted array containing numbers
 * @returns : a sorted array
 */
export default function selectionSort(arr: number[]): number[] {
    
    const copiedArr = [...arr]
    const sortedArr: number[] = []
    const initialLength = copiedArr.length
    
    // looping
    for (let i = 0; i < initialLength ; i++){
        const smallestIndex: number = findSmallest(copiedArr)
        const smallest: number = copiedArr.splice(smallestIndex, 1)[0] as number
        sortedArr.push(smallest)
    }


    return sortedArr
    
}

